src/CacheClient.cpp

```diff
@@ -0,0 +1,125 @@
+#include "CacheClient.hpp"
+#include "Hash.hpp"
+#include "DagData.hpp"
+#include "RunAction.hpp"
+#include "NodeResultPrinting.hpp"
+#include "StatCache.hpp"
+#include "Profiler.hpp"
+#include "Exec.hpp"
+#include "BuildQueue.hpp"
+#include "MakeDirectories.hpp"
+
+const char* kENV_REAPI_CACHE_CLIENT = "REAPI_CACHE_CLIENT";
+const char* kENV_CACHE_SERVER_ADDRESS = "CACHE_SERVER_ADDRESS";
+const char* kENV_BEE_CACHE_BEHAVIOUR = "BEE_CACHE_BEHAVIOUR";
+
+static int SlowCallback(void *user_data, const char* label)
+{
+    SlowCallbackData *data = (SlowCallbackData *)user_data;
+    MutexLock(data->queue_lock);
+    char buffer[1000];
+    snprintf(buffer,sizeof(buffer),"%s %s", data->node_data->m_Annotation.Get(), label);
+    int sendNextCallbackIn = PrintNodeInProgress(data->node_data, data->time_of_start, data->build_queue, buffer);
+    MutexUnlock(data->queue_lock);
+    return sendNextCallbackIn;
+}
+
+static int SlowCallback_CacheRead(void *user_data)
+{
+    return SlowCallback(user_data, "[CacheRead]");
+}
+
+static int SlowCallback_CacheWrite(void *user_data)
+{
+    return SlowCallback(user_data, "[CacheWrite]");
+}
+
+enum Operation
+{
+    kOperationRead,
+    kOperationWrite
+};
+
+static bool Invoke_REAPI_Cache_Client(const HashDigest& digest, StatCache *stat_cache, const FrozenArray<FrozenFileAndHash>& outputFiles, ThreadState* thread_state, Operation operation, const Frozen::DagNode* dagNode, Mutex* queue_lock)
+{
+    ProfilerScope profiler_scope("InvokeCacheMe", thread_state->m_ProfilerThreadId, outputFiles[0].m_Filename);
+
+    const char* reapi_raw = getenv(kENV_REAPI_CACHE_CLIENT);
+    if (reapi_raw == nullptr)
+        return false;
+
+    PathBuffer pathbuf;
+    PathInit(&pathbuf, reapi_raw);
+    char reapi[kMaxPathLength];
+    PathFormat(reapi, &pathbuf);
+
+    char buffer[5000];
+    int totalWritten = 0;
+    char digestString[kDigestStringSize];
+    DigestToString(digestString, digest);
+
+    const char* cmd = operation == kOperationRead ? "down" : "up";
+    totalWritten += snprintf(buffer, sizeof(buffer), "%s %s %s00000000000000000000000000000002", reapi, cmd, digestString);
+
+    for (auto &it : outputFiles)
+    {
+        PathBuffer output;
+        PathInit(&output, it.m_Filename);
+        MakeDirectoriesForFile(stat_cache, output);
+        totalWritten += snprintf(buffer+totalWritten, sizeof(buffer)-totalWritten, " \"%s\" ", it.m_Filename.Get());
+    }
+
+    SlowCallbackData slowCallbackData;
+    slowCallbackData.node_data = dagNode;
+    slowCallbackData.time_of_start = TimerGet();
+    slowCallbackData.queue_lock = queue_lock;
+    slowCallbackData.build_queue = thread_state->m_Queue;
+
+    Log(kDebug,"%s\n",buffer);
+    ExecResult result = ExecuteProcess(buffer, 0, nullptr, nullptr, thread_state->m_ThreadIndex, false, operation == kOperationRead ? SlowCallback_CacheRead : SlowCallback_CacheWrite , &slowCallbackData);
```

Comment:
For cache uploads, do we care about waiting for the result? Guess it would be better to just let those happen in the background, and get back to doing real work?

Reply:
would be more optimal to do in the background, but also considerably more complicated.  we can do it when we want to make it faster.

Reply:
+1 to deferring this.

(Ideally it probably makes sense to add these operations to a separate queue processed by a separate tundra thread, and make tundra not exit until that queue is also finished or timed out.)